packages/vscode-extension/src/project/project.ts

```diff
@@ -81,6 +83,39 @@ export class Project implements Disposable, MetroDelegate, DebugSessionDelegate,
     });
   }
 
+  //#region Build progress
+  onStateChange(state: StartupMessage): void {
+    this.updateProjectStateForDevice(this.projectState.selectedDevice!, { startupMessage: state });
+  }
+  //#endregion
+
+  //#region App events
+  onAppEvent<E extends keyof AppEvent, P = AppEvent[E]>(event: E, payload: P): void {
+    switch (event) {
+      case "appReady":
+        Logger.debug("App ready");
+        if (this.reloadingMetro) {
+          this.reloadingMetro = false;
+          this.updateProjectState({ status: "running" });
```

Comment:
not sure where this whole refactor is going. It seems like you're moving the logic responsible for handling selected device into device session. In that case shouldn't we also move state to live there or maybe that's what some of the follow up PRs is doing?

Reply:
Discussed offline, to sum up:
we want to move as much state from the project as we can, including current `projectState`. This can't be done in one go so I opted to do it in smaller bites.
Right now, we're establishing what events change the state so it's clearer what's happening.

As for shared state – metro, devtools – we need to think of good way of keeping them in DeviceSession that allows to create multiple DeviceSession for multiple devices. I haven't got an answer to that, yet.